server/api/species.go

```diff
@@ -6,10 +6,20 @@ import (
 	pb "github.com/backstopmedia/gRPC-book-example/server/proto"
 )
 
-func (s *Server) GetSpecies(context.Context, *pb.GetSpeciesRequest) (*pb.GetSpeciesResponse, error) {
-	return nil, s.notImplementedError()
+func (s *Server) GetSpecies(ctx context.Context, r *pb.GetSpeciesRequest) (*pb.GetSpeciesResponse, error) {
+	species, err := s.db.GetSpeciesByID(ctx, r.Id)
+	if err != nil {
+		return nil, err
+	}
```

Comment:
I think for this singular requests, we should add the functionality of returning separate codes. For example, if a request comes into this endpoint for a species that doesn't exist, it will return an arbitrary error, which in grpc world is a `code: 2`. We should add to this example the ability of switching on error types and return grpc.Errorf(codes.NotFound, ....)` for example.

Reply:
I think that's good for a separate PR though, I'd like to take a crack at that if that's cool.

Reply:
Definitely a good idea! 👍 

I've just pushed PRs for all the plumbing for the rest of the APIs. If we get them all merged (they're all the same), you'll be able to update them all with the new behavior in the same PR

Reply:
I'm tempted to do it at a server interceptor level now that I'm thinking about it.

Reply:
I kinda like that idea. We could add some specific error types to the db package to make that easier for the interceptor to determine the error type and how to respond

Reply:
PR is here: #14